### PR TITLE
Loic work

### PR DESCRIPTION
**Pull request by Ashargin**

### CODE REVIEW

File: grid2op/Opponent/NeuripsOpponent.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+
+class NeuripsOpponent(BaseOpponent):
+    def __init__(self, action_space):
+        BaseOpponent.__init__(self, action_space)
+        self._do_nothing = None
+        self._attacks = None
+        self._lines_ids = None
+        self._next_attack_time = None
+        self._attack_cooldown = 12*24 # attack cooldown
```
Comment (BDonnot) [inline grid2op/Opponent/NeuripsOpponent.py]: Non

File: grid2op/Opponent/NeuripsOpponent.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        # this is the constructor:
+        # it should have the exact same signature as here
+
+    def init(self, lines_attacked=[], **kwargs):
```
Comment (BDonnot) [inline grid2op/Opponent/NeuripsOpponent.py]: in kwargs, add the attack_cooldown if you want.

File: grid2op/Opponent/OpponentSpace.py [hunk 3]
```diff
@@ -152,7 +152,6 @@ def attack(self, observation, agent_action, env_action):
         # If currently attacking
         if self.current_attack_duration > 0:
             attack = self.last_attack
-            self.opponent.tell_attack_continues(observation, agent_action, env_action, self.budget)
```
Comment (BDonnot) [inline grid2op/Opponent/OpponentSpace.py:155]: alors je veux un test qui vérifie que :
- soit tell_attack_continues
- soit attack
est appelé à chaque time step.
Merci

**Review by BDonnot — CHANGES_REQUESTED** [review]

File: grid2op/Opponent/NeuripsOpponent.py (hunk 2 above)
Comment (Ashargin) [reply grid2op/Opponent/NeuripsOpponent.py]: done

Comment (BDonnot) [inline grid2op/Opponent/NeuripsOpponent.py]: vu ce qu'on vient de dire, essaie aussi de rajouter un vecteur w qui a la meme taille que lines_attack
et tu tireras au hasard proprotionellement à `rho / w` au lieu de tirer proportionellement à rho

**Review by BDonnot — CHANGES_REQUESTED** [review]
Test do not pass !

File: grid2op/Opponent/NeuripsOpponent.py (hunk 2 above)
Comment (Ashargin) [reply grid2op/Opponent/NeuripsOpponent.py]: done

File: grid2op/Opponent/NeuripsOpponent.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            self._attacks.append(a)
+        self._attacks = np.array(self._attacks)
+
+        # Usage rates normalization
+        self._rho_normalization = np.ones_like(lines_attacked)
+        if len(rho_normalization) == 0:
+            warnings.warn('The usage rate normalization is not specified. No normalization will be performed.')
+        elif len(rho_normalization) != len(lines_attacked):
+            warnings.warn(f'The usage rate normalization must have the same length as the number '
+                          f'of attacked lines. No normalization will be performed.')
```
Comment (BDonnot) [inline grid2op/Opponent/NeuripsOpponent.py]: this is not a warning, this is an error in this case.

File: grid2op/Opponent/NeuripsOpponent.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        else:
+            self._rho_normalization = np.array(rho_normalization)
+
+        # Opponent's attack period
+        self._attack_period = kwargs["attack_period"] if "attack_period" in kwargs else 12*24
```
Comment (BDonnot) [inline grid2op/Opponent/NeuripsOpponent.py]: then directly expand the "attack_period" to be a keyword arguments: `def init(self, lines_attacked=[], rho_normalization=[], attack_period = 12*24**kwargs):`

File: grid2op/Opponent/NeuripsOpponent.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        self._attack_period = None
+        self._rho_normalization = None
+
+        # this is the constructor:
+        # it should have the exact same signature as here
+
+    def init(self, lines_attacked=[], rho_normalization=[], **kwargs):
+        # this if the function used to properly set the object.
+        # It has the generic signature above,
+        # and it's way more flexible that the other one.
```
Comment (BDonnot) [inline grid2op/Opponent/NeuripsOpponent.py]: Have a proper documentation here.

File: grid2op/tests/test_Opponent.py [hunk 7]
```diff
@@ -9,7 +9,7 @@
 import tempfile
 import warnings
 from grid2op.tests.helper_path_test import *
-from grid2op.Opponent import BaseOpponent, RandomLineOpponent
+from grid2op.Opponent import BaseOpponent, NeuripsOpponent
```
Comment (BDonnot) [inline grid2op/tests/test_Opponent.py]: Wait i'm scared now.

Put back all the tests we did before, when we validate the opponent (for version 1.0.0) and then add some tests for NeuripsOpponent.
Don't change the tests like that.

tests are added, it's extremely rare that you removed them.

**Review by BDonnot — CHANGES_REQUESTED** [review]

File: grid2op/Opponent/NeuripsOpponent.py (hunk 4 above)
Comment (Ashargin) [reply grid2op/Opponent/NeuripsOpponent.py]: done

File: grid2op/Opponent/NeuripsOpponent.py (hunk 5 above)
Comment (Ashargin) [reply grid2op/Opponent/NeuripsOpponent.py]: done

File: grid2op/Opponent/NeuripsOpponent.py (hunk 6 above)
Comment (Ashargin) [reply grid2op/Opponent/NeuripsOpponent.py]: done

File: grid2op/tests/test_Opponent.py (hunk 7 above)
Comment (Ashargin) [reply grid2op/tests/test_Opponent.py]: done

File: grid2op/Opponent/OpponentSpace.py (hunk 3 above)
Comment (Ashargin) [reply grid2op/Opponent/OpponentSpace.py:155]: done

**Review by BDonnot — APPROVED** [review]
I'll check tomorrow with a diff between the test_Opponent of master and this one that for now is a bit "messy" in here, but beside that it looks great.

File: grid2op/Opponent/NeuripsOpponent.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+# This file is part of Grid2Op, Grid2Op a testbed platform to model sequential decision making in power systems.
+import warnings
+import numpy as np
+
+from grid2op.Opponent import BaseOpponent
+from grid2op.Exceptions import OpponentError
+
+
+class NeuripsOpponent(BaseOpponent):
+    def __init__(self, action_space):
```
Comment (BDonnot) [inline grid2op/Opponent/NeuripsOpponent.py]: short description here

File: grid2op/Opponent/NeuripsOpponent.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    def init(self, lines_attacked=[], rho_normalization=[], attack_period=12*24, **kwargs):
+        """
+        This function initializes the parameters of the NeuripsOpponent properly.
+        lines_attacked : list: the list of lines that the NeuripsOpponent should be able to disconnect
+        rho_normalization: list: the list of mean usage rates for the attackable lines. Should have
+                                 the same length as lines_attacked. If no value is given, no normalization
+                                 will be performed.
+        attack_period: int: the number of steps among which the attack may happen.
+                            If attack_period=10, then when an attack can be made, it will happen in the 10
+                            next steps.
```
Comment (BDonnot) [inline grid2op/Opponent/NeuripsOpponent.py]: please follow the doc format.

**Review by BDonnot — COMMENTED** [review]
Improve also the doc if possible. People will probably have a look at this opponent.

File: grid2op/Opponent/NeuripsOpponent.py (hunk 8 above)
Comment (Ashargin) [reply grid2op/Opponent/NeuripsOpponent.py]: done

File: grid2op/Opponent/NeuripsOpponent.py (hunk 9 above)
Comment (Ashargin) [reply grid2op/Opponent/NeuripsOpponent.py]: done

File: grid2op/Opponent/NeuripsOpponent.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+# SPDX-License-Identifier: MPL-2.0
+# This file is part of Grid2Op, Grid2Op a testbed platform to model sequential decision making in power systems.
+import warnings
+import numpy as np
+
+from grid2op.Opponent import BaseOpponent
+from grid2op.Exceptions import OpponentError
+
+
+class NeuripsOpponent(BaseOpponent):
```
Comment (BDonnot) [inline grid2op/Opponent/NeuripsOpponent.py]: Il n'a pas changé de nom :-(

File: grid2op/Opponent/NeuripsOpponent.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        """
+        This function initializes the parameters of the NeuripsOpponent properly.
+        lines_attacked : list: the list of lines that the NeuripsOpponent should be able to disconnect
+        rho_normalization: list: the list of mean usage rates for the attackable lines. Should have
+                                 the same length as lines_attacked. If no value is given, no normalization
+                                 will be performed.
+        attack_period: int: the number of steps among which the attack may happen.
+                            If attack_period=10, then when an attack can be made, it will happen in the 10
+                            next steps.
+        """
```
Comment (BDonnot) [inline grid2op/Opponent/NeuripsOpponent.py]: Doc toujours pas au bon format

File: grid2op/Opponent/NeuripsOpponent.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        self._next_attack_time -= 1
+
+        # If the attack time has not come yet, do not attack
+        if self._next_attack_time > 0:
+            return None
+
+        # If all attackable lines are disconnected, do not attack
+        status = observation.line_status[self._lines_ids]
+        if np.all(status == False):
+            return None
```
Comment (BDonnot) [inline grid2op/Opponent/NeuripsOpponent.py]: re sample the _next_attac_time in this case, no ?

**Review by BDonnot — CHANGES_REQUESTED** [review]
Change the name of the opponent and update the doc please :-)

Beside that it looks good.

File: grid2op/Opponent/NeuripsOpponent.py (hunk 10 above)
Comment (Ashargin) [reply grid2op/Opponent/NeuripsOpponent.py]: Je pense que si?

File: grid2op/Opponent/NeuripsOpponent.py (hunk 11 above)
Comment (Ashargin) [reply grid2op/Opponent/NeuripsOpponent.py]: Si c'est bon tu as du commenté avant mon dernier commit

File: grid2op/Opponent/NeuripsOpponent.py (hunk 12 above)
Comment (Ashargin) [reply grid2op/Opponent/NeuripsOpponent.py]: I don't think so. In that case where it is time to attack but all lines are disconnected, the opponent will simply attack as soon as a line is available to attack (connected).

**Review by BDonnot — APPROVED** [review]

**Pull request merged**